File: src/game/keybinds.ts
// Rebindable keybinds system.
// Stores a mapping from canonical game action -> keyboard key (lowercased KeyboardEvent.key).
// Persists to localStorage under POTC_KEYBINDS_V1.

export type GameAction =
  | "moveLeft" | "moveRight" | "moveUp" | "moveDown"
  | "jump" | "dash" | "roll" | "parry"
  | "fire" | "melee" | "miscA" | "miscB" | "grab"
  | "shield" | "overdrive"
  | "inventory" | "pause"
  | "slot1" | "slot2" | "slot3" | "slot4" | "slot5" | "slot6";

export const ALL_ACTIONS: GameAction[] = [
  "moveLeft","moveRight","moveUp","moveDown",
  "jump","dash","roll","parry",
  "fire","melee","miscA","miscB","grab",
  "shield","overdrive",
  "inventory","pause",
  "slot1","slot2","slot3","slot4","slot5","slot6",
];

export const ACTION_LABEL: Record<GameAction, string> = {
  moveLeft: "Move Left", moveRight: "Move Right", moveUp: "Up / Climb Up", moveDown: "Down / Climb Down",
  jump: "Jump (2nd bind)", dash: "Dash", roll: "Roll", parry: "Parry",
  fire: "Fire Ranged", melee: "Melee", miscA: "Misc A (hold to charge)", miscB: "Misc B (hold to charge)", grab: "Grab (hold to charge throw)",
  shield: "Shield", overdrive: "Overdrive",
  inventory: "Inventory", pause: "Pause",
  slot1: "Ranged Slot 1", slot2: "Slot 2", slot3: "Slot 3", slot4: "Slot 4", slot5: "Slot 5", slot6: "Slot 6",
};

export const DEFAULT_KEYBINDS: Record<GameAction, string> = {
  moveLeft: "a", moveRight: "d", moveUp: "w", moveDown: "s",
  jump: " ",     // space (W also jumps in engine special-case)
  dash: "q",
  roll: "z",
  parry: "e",
  fire: "f",
  melee: "r",
  miscA: "o",
  miscB: "p",
  grab: "v",
  shield: "x",
  overdrive: "g",
  inventory: "tab",
  pause: "escape",
  slot1: "1", slot2: "2", slot3: "3", slot4: "4", slot5: "5", slot6: "6",
};

const STORAGE_KEY = "POTC_KEYBINDS_V1";

export function loadKeybinds(): Record<GameAction, string> {
  try {
    const raw = localStorage.getItem(STORAGE_KEY);
    if (!raw) return { ...DEFAULT_KEYBINDS };
    const saved = JSON.parse(raw) as Partial<Record<GameAction, string>>;
    return { ...DEFAULT_KEYBINDS, ...saved };
  } catch {
    return { ...DEFAULT_KEYBINDS };
  }
}

export function saveKeybinds(kb: Record<GameAction, string>) {
  try { localStorage.setItem(STORAGE_KEY, JSON.stringify(kb)); } catch {}
}

export function resetKeybinds(): Record<GameAction, string> {
  try { localStorage.removeItem(STORAGE_KEY); } catch {}
  return { ...DEFAULT_KEYBINDS };
}

// Normalize a KeyboardEvent.key into the lowercase canonical form we store.
export function normalizeKey(k: string): string {
  const low = k.toLowerCase();
  if (low === "spacebar") return " ";
  return low;
}

// Pretty-print a stored key for display.
export function prettyKey(k: string): string {
  if (k === " ") return "SPACE";
  if (k === "escape") return "ESC";
  if (k === "tab") return "TAB";
  if (k === "arrowleft") return "←";
  if (k === "arrowright") return "→";
  if (k === "arrowup") return "↑";
  if (k === "arrowdown") return "↓";
  return k.toUpperCase();
}

// Singleton accessor used by the engine
let current: Record<GameAction, string> = loadKeybinds();
export function getKeybinds() { return current; }
export function setKeybinds(kb: Record<GameAction, string>) { current = { ...kb }; saveKeybinds(current); }

// Reverse lookup: which action does this key trigger?
export function actionFor(key: string): GameAction | null {
  const k = normalizeKey(key);
  for (const a of ALL_ACTIONS) if (current[a] === k) return a;
  return null;
}

File: python_recreation/pygame_ce_port/data/keybinds.py
"""data/keybinds.py — load keybinds.json, normalize to pygame keycodes."""
import json, os
import pygame

DEFAULT_KEYBINDS = {
    "moveLeft": "a", "moveRight": "d", "moveUp": "w", "moveDown": "s",
    "jump": "space", "dash": "q", "roll": "z", "parry": "e",
    "fire": "f", "melee": "r", "miscA": "o", "miscB": "p",
    "grab": "v", "shield": "x", "overdrive": "g",
    "inventory": "tab", "pause": "escape", "shop": "t", "interact": "return",
    "slot1": "1", "slot2": "2", "slot3": "3", "slot4": "4", "slot5": "5", "slot6": "6",
}

_ALIASES = {
    "space": pygame.K_SPACE, "escape": pygame.K_ESCAPE, "tab": pygame.K_TAB,
    "return": pygame.K_RETURN, "enter": pygame.K_RETURN,
    "shift": pygame.K_LSHIFT, "ctrl": pygame.K_LCTRL, "alt": pygame.K_LALT,
    "left": pygame.K_LEFT, "right": pygame.K_RIGHT, "up": pygame.K_UP, "down": pygame.K_DOWN,
}

def key_to_pg(name: str) -> int:
    name = (name or "").lower().strip()
    if name in _ALIASES: return _ALIASES[name]
    if not name: return pygame.K_UNKNOWN
    try: return pygame.key.key_code(name)
    except Exception: return pygame.K_UNKNOWN

def load_keybinds() -> dict:
    here = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    path = os.path.join(here, "keybinds.json")
    binds = dict(DEFAULT_KEYBINDS)
    if os.path.exists(path):
        try:
            with open(path) as f: binds.update(json.load(f))
        except Exception: pass
    return binds

def save_keybinds(binds: dict) -> None:
    here = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    path = os.path.join(here, "keybinds.json")
    try:
        with open(path, "w") as f: json.dump(binds, f, indent=2)
    except Exception: pass
